rfcs/2021-05-05-7351-framing-and-codecs.md

```diff
@@ -0,0 +1,183 @@
+uniform way. It does not cover any specific implementation for framing or a
+codec. Also see [future work](#future-work) for extended goals that are not
+covered by this RFC.
+
+## Motivation
+
+Currently, we have no explicit abstraction that is responsible for handling
+framing and codecs, such that each source and sink may implement them in their
+own way. This can introduce inconsistencies and additional maintenance burden.
+Components have their own defaults, making behavior unpredictable and surprising
+users, e.g. as documented in
+[#3453](https://github.com/timberio/vector/issues/3453).
+
+## Internal Proposal
+
+Since decoding and encoding should apply universally to all sources and sinks,
+it should be configurable by a shared field. This is accomplished by the
+`SourceOuter` and `SinkOuter` wrappers:
+
+```rust
+// Newly introduced wrapper, analogously to `SinkOuter`.
+struct SourceOuter {
+    framing: Framer,
+    decoding: Decoder,
+}
+```
+
+```rust
+struct SinkOuter {
+    // … existing fields
+    encoding: Encoder,
+}
+```
+
+Conceptually, decoding is applied _after_ a source has processed an event and
+encoding is applied _before_ a sink has processed an event. These can be
+implemented as `FunctionTransform`s in our pipeline, in the same way transforms
+are implemented.
+
+Framing takes place before an event is created since the message boundary
+required to form an event is not known yet. Therefore the source needs to be
+aware of framing and call the framer to determine message boundaries before
+creating an event.
+
+To make the source aware of framing, the `Framer` needs to be passed to the
+`SourceContext`, so that the source implementation can call it:
+
+```rust
+struct SourceContext {
+    // … existing fields
+    framing: Framer,
+}
+```
+
+In the context of Vector, decoders and encoders implement functions with the
+signature `fn decode(value: Value) -> Result<Value>` /
+`fn encode(value: Value) -> Result<Value>` where e.g. a `json` encoder can
+convert a `Value::Bytes` to `Value::Map`.
+
+For the implementation of a framer, we can defer to Tokio's
+[`FramedRead`](https://docs.rs/tokio-util/0.6.6/tokio_util/codec/struct.FramedRead.html)
+trait which does exactly what we want here.
+
+## Doc-level Proposal
+
+// TODO.
+
+<!--
+- Optional. Only do this if your change is public facing.
+- Demonstrate how your change will look in the form of Vector's public docs.
+-->
+
+## Rationale
+
+One prime example where a source's implementation may be reused with a different
+codec is the `syslog` source (see
+[#7046](https://github.com/timberio/vector/pull/7046)), or the upcoming `syslog`
+sink in [#7106](https://github.com/timberio/vector/issues/7106). Instead of
+re-implementing socket-based connection handling, the `syslog` components could
+be replaced by the `socket` counterparts combined with `octet-framing`. This
+reduces a possible source of bugs and inconsistencies and therefore leads to
+less maintenance burden.
```

Comment:
Syslog is actually not the best example, because I think we would still want to expose a first-class `syslog` source to users. We could expand that config using codecs internally, but that's not quite the main purpose here.

Maybe a better motivating example would be something like reading JSON-encoded from Kafka. Right now that requires configuring and wiring together two separate components, where with codecs it could be a simple one-line change to the config.

Reply:
JSON decoding events as they come into the `kafka` source has come up pretty frequently in the community!

Reply:
Thanks, makes sense. Added that as main motivation 40672fc505a0a436aa383b2c490eac205450fbc2